more_itertools/more.py

```diff
@@ -1654,35 +1654,145 @@ def numeric_range(*args):
+            self._start, self._stop = args
+            self._step = 1
+        elif argc == 3:
+            self._start, self._stop, self._step = args
+        elif argc == 0:
+            raise TypeError('numeric_range expected at least '
+                            '1 argument, got {}'.format(argc))
+        else:
+            raise TypeError('numeric_range expected at most '
+                            '3 arguments, got {}'.format(argc))
+
+        self._zero = type(self._step)(0)
+        if self._step == self._zero:
+            raise ValueError('numeric_range() arg 3 must not be zero')
+        self._growing = self._step > self._zero
+
+    def __bool__(self):
+        if self._growing:
+            return self._start < self._stop
+        else:
+            return self._start > self._stop
+
+    def __contains__(self, elem):
+        if self._growing:
+            if self._start <= elem < self._stop:
+                return (elem - self._start) % self._step == self._zero
+        else:
+            if self._start >= elem > self._stop:
+                return (self._start - elem) % (-self._step) == self._zero
+
+        return False
+
+    def __eq__(self, other):
+        return (self._start == other._start
+                and self._step == other._step
+                and self._get_by_index(-1) == other._get_by_index(-1))
+
+    def __getitem__(self, key):
+        if isinstance(key, int):
+            return self._get_by_index(key)
+        elif isinstance(key, slice):
+            start = self._start if key.start is None \
+                else self._get_by_index(key.start)
+            stop = self._stop if key.stop is None \
+                else self._get_by_index(key.stop)
+            step = self._step if key.step is None else key.step * self._step
+            return numeric_range(start, stop, step)
+        else:
+            raise TypeError(
+                'numeric range indices must be '
+                'integers or slices, not {}'.format(type(key).__name__))
+
+    def __hash__(self):
+        if self:
+            return hash((self._start, self._get_by_index(-1), self._step))
+        else:
+            return self._EMPTY_HASH
+
+    def __iter__(self):
+        values = (self._start + (self._step * n) for n in count())
+        if self._growing:
+            return takewhile(partial(gt, self._stop), values)
+        else:
+            return takewhile(partial(lt, self._stop), values)
+
+    def __len__(self):
+        if self._growing:
+            start = self._start
+            stop = self._stop
+            step = self._step
+        else:
+            start = self._stop
+            stop = self._start
+            step = -self._step
+
+        distance = stop - start
+        if distance <= 0:
+            return 0
+
+        # distance > 0 and step > 0: regular euclidean division
+        q, r = divmod(distance, step)
+        return int(q) + int(r != self._zero)
+
+    def __reduce__(self):
```

Comment:
This method doesn't appear to be tested - could we add a pickle test?

Reply:
Done.